### PR TITLE
chore: extract flagd container injection into its own component

### PR DESCRIPTION
**Pull request by bacherfl**
This PR refactors the Pod mutating webhook by extracting the functionality of the flagd container injection, as well as the setup of required permissions and configmaps containing the feature flag configuration into its own component.

The motivation for this is that the flagd container injection will also be used in the `Flagd` CRD's controller when [OFO Flag Service OFEP](https://github.com/open-feature/ofep/blob/main/OFEP-ofo-flag-service.md) is going to be implemented.

In addition to the migration of the flagd container injection functionality and the tests thereof, this PR also introduces more detailed unit tests that cover the handling of different sync source providers

### CODE REVIEW

File: controllers/common/flagd-injector.go [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+	objectMeta.Annotations[fmt.Sprintf("%s/%s", constant.OpenFeatureAnnotationPrefix, constant.AllowKubernetesSyncAnnotation)] = "true"
+
+	// build K8s config
+	return types.SourceConfig{
+		URI:      fmt.Sprintf("%s/%s", ns, n),
+		Provider: string(v1alpha1.SyncProviderKubernetes),
+	}, nil
+}
+
+func (fi *FlagdContainerInjector) getFeatureFlag(ctx context.Context, namespace string, name string) v1alpha1.FeatureFlagConfiguration {
```
Comment (Kavindu-Dodan) [inline controllers/common/flagd-injector.go:347]: Seems to be a duplicate of common `func FeatureFlag(ctx context.Context, c client.Client, namespace string, name string)`

File: controllers/common/flagd-injector.go [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+	ffConfig := v1alpha1.FeatureFlagConfiguration{}
+	// try to retrieve the FeatureFlagConfiguration
+	if err := fi.Client.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, &ffConfig); errors.IsNotFound(err) {
+		return v1alpha1.FeatureFlagConfiguration{}
+	}
+	return ffConfig
+}
+
+func (fi *FlagdContainerInjector) appendSources(sources []types.SourceConfig, sidecar *corev1.Container) error {
```
Comment (Kavindu-Dodan) [inline controllers/common/flagd-injector.go:356]: can remove the pointer receiver and make a simple helper function

File: controllers/common/flagd-injector.go [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+				Name:          "metrics",
+				ContainerPort: flagSourceConfig.MetricsPort,
+			},
+		},
+		SecurityContext: getSecurityContext(),
+		Resources:       fi.FlagDResourceRequirements,
+	}
+}
+
+func (fi *FlagdContainerInjector) addFlagdContainer(spec *corev1.PodSpec, flagdContainer corev1.Container) {
```
Comment (Kavindu-Dodan) [inline controllers/common/flagd-injector.go:391]: can remove the pointer receiver and made to a simple helper function

**Comment by Kavindu-Dodan** [comment]
Overal great work 👍  A refactoring was due for these code sections and good that you worked on it.

File: pkg/constant/configuration.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package constant
```
Comment (Kavindu-Dodan) [inline pkg/constant/configuration.go:1]: minor - I see the packaging and location of the constant little odd. Should we move this to controller ? Or was there any reason to introduce it here :) ?

**Comment by bacherfl** [comment]
> Overal great work 👍 A refactoring was due for these code sections and good that you worked on it.

Thanks for the suggestions @Kavindu-Dodan ! I agree with all of them, and have updated the PR accordingly

**Review by Kavindu-Dodan — APPROVED** [review]

**Review by thisthat — APPROVED** [review]
LGTM

**Pull request merged**